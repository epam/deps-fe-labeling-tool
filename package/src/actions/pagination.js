import { createAction } from 'redux-actions'

const FEATURE_NAME = 'PAGINATION'

const openPage = createAction(`${FEATURE_NAME}/OPEN_PAGE`)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

export {
  resetDefault,
  openPage
}
