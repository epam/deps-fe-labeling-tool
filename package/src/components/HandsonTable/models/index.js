
export * from './HTTableData'
export * from './HTMerge'
export * from './HTColumn'
