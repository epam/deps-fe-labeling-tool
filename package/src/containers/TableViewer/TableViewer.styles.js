
import styled from 'styled-components'

export const TableViewerContainer = styled.div`
  height: 100%;
  display: flex;
  flex-direction: column;
  align-content: center;
  padding: 5rem 1rem;
`
