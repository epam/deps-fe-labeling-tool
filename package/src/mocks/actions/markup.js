const addLabels = jest.fn((labels) => ({
  type: 'MOCK_ADD_LABELS_ACTION',
  payload: labels
}))

const addTables = jest.fn((tables) => ({
  type: 'MOCK_ADD_TABLES_ACTION',
  payload: tables
}))

const removeLabels = jest.fn((labels) => ({
  type: 'MOCK_REMOVE_LABELS_ACTION',
  payload: labels
}))

const updateLabels = jest.fn((labels) => ({
  type: 'MOCK_UPDATE_LABELS_ACTION',
  payload: labels
}))

const updateAllLabels = jest.fn((labels) => ({
  type: 'MOCK_UPDATE_ALL_LABELS_ACTION',
  payload: labels
}))

const updateTables = jest.fn((tables) => ({
  type: 'MOCK_UPDATE_TABLES_ACTION',
  payload: tables
}))

const updateAllTables = jest.fn((tables) => ({
  type: 'MOCK_UPDATE_ALL_TABLES_ACTION',
  payload: tables
}))

const selectLabels = jest.fn((labels) => ({
  type: 'MOCK_SELECT_LABELS_ACTION',
  payload: labels
}))

const selectTables = jest.fn((tables) => ({
  type: 'MOCK_SELECT_TABLES_ACTION',
  payload: tables
}))

const clearSelection = jest.fn(() => ({
  type: 'MOCK_CLEAR_SELECTION_ACTION'
}))

const removeTables = jest.fn((tables) => ({
  type: 'MOCK_REMOVE_TABLES_ACTION',
  payload: tables
}))

const importMarkup = jest.fn(() => ({
  type: 'MOCK_IMPORT_MARKUP'
}))

const exportMarkup = jest.fn(() => ({
  type: 'MOCK_EXPORT_MARKUP'
}))

const undo = jest.fn(() => ({
  type: 'MOCK_UNDO_ACTION'
}))

const redo = jest.fn(() => ({
  type: 'MOCK_REDO_ACTION'
}))

const updateInitialMarkup = jest.fn(() => ({
  type: 'MOCK_UPDATE_INITIAL_MARKUP_ACTION'
}))

const storeMarkup = jest.fn((markup) => ({
  type: 'MOCK_STORE_MARKUP_ACTION',
  payload: markup
}))

const storeAssignedMarkup = jest.fn((markup) => ({
  type: 'MOCK_UPDATE_MARKUP_ACTION',
  payload: markup
}))

const updatePageMarkup = jest.fn(() => ({
  type: 'UPDATE_PAGE_MARKUP'
}))

const reset = jest.fn(() => ({
  type: 'RESET_MARKUP_ACTION'
}))

const copyMarkup = jest.fn(() => ({
  type: 'COPY_MARKUP_ACTION'
}))

const pasteMarkup = jest.fn(() => ({
  type: 'PASTE_MARKUP_ACTION'
}))

const updateLabelsWithSettings = jest.fn(() => ({
  type: 'UPDATE_LABELS_WITH_SETTINGS'
}))

const updateTablesWithSettings = jest.fn(() => ({
  type: 'UPDATE_TABLES_WITH_SETTINGS'
}))

const mockMarkupActions = {
  addLabels,
  addTables,
  copyMarkup,
  pasteMarkup,
  removeLabels,
  reset,
  updateLabels,
  updateAllLabels,
  updateTables,
  updateAllTables,
  selectLabels,
  selectTables,
  clearSelection,
  removeTables,
  importMarkup,
  exportMarkup,
  undo,
  updateInitialMarkup,
  storeMarkup,
  storeAssignedMarkup,
  updatePageMarkup,
  redo,
  updateLabelsWithSettings,
  updateTablesWithSettings
}

export {
  mockMarkupActions
}
