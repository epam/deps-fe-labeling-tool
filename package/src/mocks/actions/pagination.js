const openPage = jest.fn((page) => ({
  type: 'MOCK_OPEN_PAGE_ACTION',
  payload: page
}))

const mockPaginationActions = {
  openPage
}

export {
  mockPaginationActions
}
