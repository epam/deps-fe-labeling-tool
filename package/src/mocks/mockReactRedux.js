const mockDispatch = jest.fn((action) => action)

const createMockMapStateToProps = (mapStateToProps) => {
  if (!mapStateToProps) {
    return mapStateToProps
  }

  return (state, ownProps) => ({
    props: mapStateToProps(state, ownProps)
  })
}

const getDispatchFunctionFormShorthandObject = (actionsObj) => (dispatch) => (
  Object.keys(actionsObj).reduce(
    (props, actionKey) => ({
      ...props,
      [actionKey]: jest.fn(
        (...args) => dispatch(actionsObj[actionKey].apply(undefined, args))
      )
    }),
    {}
  )
)

const createMockMapDispatchToProps = (mapDispatchToProps) => {
  if (!mapDispatchToProps) {
    return mapDispatchToProps
  }

  const mappingFunction = typeof mapDispatchToProps === 'function'
    ? mapDispatchToProps
    : getDispatchFunctionFormShorthandObject(mapDispatchToProps)

  return (dispatch, ownProps) => {
    dispatch = dispatch || mockDispatch

    return {
      props: mappingFunction(dispatch, ownProps),
      dispatch
    }
  }
}

const createMockMergeProps = (mergeProps) => {
  if (!mergeProps) {
    return mergeProps
  }

  return (stateProps, dispatchProps, ownProps) => {
    dispatchProps = dispatchProps || { dispatch: mockDispatch }

    return {
      props: mergeProps(stateProps, dispatchProps, ownProps),
      dispatch: dispatchProps.dispatch
    }
  }
}

const mockReactRedux = {
  connect: (mapStateToProps, mapDispatchToProps, mergeProps) => (WrappedComponent) => ({
    mapStateToProps: createMockMapStateToProps(mapStateToProps),
    mapDispatchToProps: createMockMapDispatchToProps(mapDispatchToProps),
    mergeProps: createMockMergeProps(mergeProps),
    WrappedComponent
  }),
  batch: (fn) => Promise.resolve(fn),
  useDispatch: jest.fn(() => mockDispatch),
  useSelector: jest.fn((selector) => selector())
}

export {
  mockReactRedux
}
