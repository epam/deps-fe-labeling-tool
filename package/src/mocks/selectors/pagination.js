const currentPageSelector = jest.fn(() => 1)

const mockPaginationSelectors = {
  currentPageSelector
}

export {
  mockPaginationSelectors
}
