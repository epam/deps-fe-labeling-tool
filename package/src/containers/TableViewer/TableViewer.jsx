
import React, { useCallback } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { updateTables } from '@/actions/markup'
import { HandsonTable } from '@/components/HandsonTable'
import { tableShape } from '@/models/Table'
import { mapMarkupTableToHandsonDataStrings } from '@/models/Table/mappers'
import { pageSelectedTablesSelector } from '@/selectors/markup'
import { currentPageSelector } from '@/selectors/pagination'
import { TableViewerContainer } from './TableViewer.styles'

const TableViewer = ({ updateTables, currentPage, selectedTables: [table] }) => {
  const saveData = useCallback(([row, column, prevValue, newValue]) => {
    const newValues = table.values.map((value) => {
      if (value.row === row && value.column === column) {
        return {
          ...value, value: newValue
        }
      }

      return value
    })

    updateTables(currentPage, [{ ...table, values: newValues }])
  }, [currentPage, table, updateTables])

  const renderTable = useCallback(() => {
    const { HTDataForRender, mergeCells } = mapMarkupTableToHandsonDataStrings(table)

    return (
      <HandsonTable
        data={HTDataForRender}
        mergeCells={mergeCells}
        saveData={saveData}
      />
    )
  }, [saveData, table])

  return (
    <TableViewerContainer>
      {renderTable()}
    </TableViewerContainer>

  )
}

TableViewer.propTypes = {
  selectedTables: PropTypes.arrayOf(tableShape).isRequired,
  updateTables: PropTypes.func.isRequired,
  currentPage: PropTypes.number.isRequired
}

const mapStateToProps = (state) => ({
  currentPage: currentPageSelector(state),
  selectedTables: pageSelectedTablesSelector(state)
})

const mapDispatchToProps = {
  updateTables
}

const ConnectedComponent = connect(mapStateToProps, mapDispatchToProps)(TableViewer)

export {
  ConnectedComponent as TableViewer
}
