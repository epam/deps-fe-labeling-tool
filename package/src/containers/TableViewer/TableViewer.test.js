
import { mockMarkupActions } from '@/mocks/actions/markup'
import { mockReactRedux } from '@/mocks/mockReactRedux'
import { mockTables } from '@/mocks/selectors/markup'
import { mockPaginationSelectors } from '@/mocks/selectors/pagination'
import React from 'react'
import { shallow } from 'enzyme'
import { updateTables } from '@/actions/markup'
import { CellValue } from '@/models/Table'
import { TableViewer } from './TableViewer'

jest.mock('@/actions/markup', () => mockMarkupActions)
jest.mock('@/selectors/pagination', () => mockPaginationSelectors)
jest.mock('@/selectors/markup', () => ({
  pageSelectedTablesSelector: jest.fn(() => mockTables)
}))
jest.mock('react-redux', () => mockReactRedux)
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useCallback: jest.fn((f) => f)
}))

const mockCellValue = [0, 0, 'Test content']

const mockNewCellData = [...mockCellValue, 'new Value']

const {
  WrappedComponent,
  mapStateToProps,
  mapDispatchToProps
} = TableViewer

describe('Component: TableViewer', () => {
  let defaultProps
  let wrapper

  beforeEach(() => {
    defaultProps = {
      ...mapStateToProps().props,
      ...mapDispatchToProps().props
    }

    wrapper = shallow(<WrappedComponent {...defaultProps} />)
  })

  it('should render TableViewer with correct props', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should trigger updateTables with correct props when call saveData', () => {
    wrapper.props().children.props.saveData(mockNewCellData)
    expect(updateTables).nthCalledWith(1, defaultProps.currentPage, [{ ...defaultProps.selectedTables[0], values: [new CellValue(0, 0, 'new Value', 0.8)] }])
  })
})
