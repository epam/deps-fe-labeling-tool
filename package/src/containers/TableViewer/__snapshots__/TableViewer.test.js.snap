// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: TableViewer should render TableViewer with correct props 1`] = `
<TableViewerstyles__TableViewerContainer>
  <Memo(HandsonTable)
    data={
      Array [
        Array [
          "value",
          null,
        ],
        Array [
          null,
          null,
        ],
      ]
    }
    mergeCells={Array []}
    saveData={[Function]}
  />
</TableViewerstyles__TableViewerContainer>
`;
